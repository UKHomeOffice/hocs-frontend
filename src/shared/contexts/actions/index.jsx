import types from './types.jsx';

export function updateForm(form) {
    return {
        type: types.UPDATE_FORM,
        payload: form
    };
}

export function updateFormData(data) {
    return {
        type: types.UPDATE_FORM_DATA,
        payload: data
    };
}

export function updateCaseSummary(data) {
    return {
        type: types.UPDATE_CASE_SUMMARY,
        payload: data
    };
}

export function updateCaseConfig(data) {
    return {
        type: types.UPDATE_CASE_CONFIG,
        payload: data
    };
}

export function updateCaseData(data) {
    return {
        type: types.UPDATE_CASE_DATA,
        payload: data
    };
}

export function unsetCaseData() {
    return {
        type: types.UPDATE_CASE_DATA,
        payload: null
    };
}

export function updateCaseActionData(data) {
    return {
        type: types.UPDATE_CASE_ACTION_DATA,
        payload: data
    };
}

export function unSetCaseActionData() {
    return {
        type: types.UNSET_CASE_ACTION_DATA,
        payload: null
    };
}

export function updateFormErrors(data) {
    return {
        type: types.UPDATE_FORM_ERRORS,
        payload: data
    };
}

export function setError(error) {
    return {
        type: types.SET_ERROR,
        payload: error
    };
}

export function unsetError() {
    return {
        type: types.UNSET_ERROR,
    };
}

export function unsetForm() {
    return {
        type: types.UNSET_FORM,
    };
}

export function updateLocation(location) {
    return {
        type: types.UPDATE_LOCATION,
        payload: location
    };
}

export function cancel() {
    return {
        type: types.CANCEL
    };
}

export function redirected() {
    return {
        type: types.REDIRECTED
    };
}

export function redirect(url) {
    return {
        type: types.REDIRECT,
        payload: url
    };
}

export function updateApiStatus(status) {
    return {
        type: types.UPDATE_API_STATUS,
        payload: {
            status,
            timeStamp: Date.now()
        }
    };
}

export function clearApiStatus() {
    return {
        type: types.CLEAR_API_STATUS
    };
}

export function updatePageMeta(meta) {
    return {
        type: types.UPDATE_PAGE_META,
        payload: meta
    };
}

export function updateDashboard(data) {
    return {
        type: types.UPDATE_DASHBOARD,
        payload: data
    };
}

export function clearDashboard() {
    return {
        type: types.CLEAR_DASHBOARD,
        payload: null
    };
}

export function updateWorkstack(data) {
    return {
        type: types.UPDATE_WORKSTACK,
        payload: data
    };
}

export function clearWorkstack() {
    return {
        type: types.CLEAR_WORKSTACK,
        payload: null
    };
}

export function unsetCaseSummary() {
    return {
        type: types.UNSET_CASE_SUMMARY,
        payload: null
    };
}

export function unsetCorrespondents() {
    return {
        type: types.UNSET_CORRESPONDENTS,
        payload: null
    };
}

export function unsetCaseNotes() {
    return {
        type: types.UNSET_CASE_NOTES,
        payload: null
    };
}

export function unsetCaseConfig() {
    return {
        type: types.UNSET_CASE_CONFIG,
        payload: null
    };
}

export function passForwardProps(props) {
    return {
        type: types.PASS_FORWARD_PROPS,
        payload: props
    };
}

export function unsetDocuments() {
    return {
        type: types.UNSET_DOCUMENTS
    };
}
