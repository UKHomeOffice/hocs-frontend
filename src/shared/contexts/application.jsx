import React, { Component } from 'react';
import PropTypes from 'prop-types';
import types from './actions/types.jsx';

const Context = React.createContext();

const reducer = (state, action) => {
    // TODO: REMOVE
    /* eslint-disable-next-line  no-console*/
    console.log(`ACTION: ${action.type} PAYLOAD: ${JSON.stringify(action.payload)}`);
    // ------------
    switch (action.type) {
    case types.UPDATE_FORM:
        return {
            ...state, form: action.payload
        };
    case types.UPDATE_FORM_DATA:

        return {
            ...state, form: {
                ...state.form, data: state.form && state.form.data ? { ...state.form.data, ...action.payload } : { ...action.payload }
            }
        };
    case types.UPDATE_FORM_ERRORS:
        return {
            ...state, form: {
                ...state.form, errors: action.payload
            }
        };
    case types.UPDATE_LOCATION:
        return { ...state, location: action.payload };
    case types.CANCEL:
        return { ...state, redirect: '/', form: null };
    case types.REDIRECT:
        return { ...state, redirect: action.payload, form: null };
    case types.REDIRECTED:
        return { ...state, redirect: null };
    case types.SET_ERROR:
        return { ...state, error: action.payload };
    case types.UNSET_ERROR:
        return { ...state, error: null };
    case types.UNSET_FORM:
        return { ...state, form: null };
    case types.UPDATE_API_STATUS:
        return { ...state, apiStatus: action.payload };
    case types.CLEAR_API_STATUS:
        return { ...state, apiStatus: null };
    case types.UPDATE_PAGE_META:
        return { ...state, page: { ...action.payload } };
    case types.UPDATE_DASHBOARD:
        return { ...state, dashboard: action.payload };
    case types.CLEAR_DASHBOARD:
        return { ...state, dashboard: null };
    case types.UPDATE_WORKSTACK:
        return { ...state, workstack: action.payload };
    case types.CLEAR_WORKSTACK:
        return { ...state, workstack: null };
    default:
        // TODO: Remove
        /* eslint-disable-next-line  no-console*/
        console.warn('Unsupported action');
    }
};

export class ApplicationProvider extends Component {
    constructor(props) {
        super(props);
        this.state = {
            ...props.config,
            apiStatus: null,
            dispatch: action => {
                try {
                    this.setState(state => reducer(state, action));
                    return Promise.resolve();
                } catch (error) {
                    return Promise.reject(error);
                }
            }
        };
    }

    render() {
        const { state, props: { children } } = this;
        return (
            <Context.Provider value={state}>
                {children}
            </Context.Provider>
        );
    }

}

ApplicationProvider.propTypes = {
    children: PropTypes.node,
    config: PropTypes.object
};

export const ApplicationConsumer = Context.Consumer;