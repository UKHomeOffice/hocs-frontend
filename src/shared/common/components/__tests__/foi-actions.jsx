import React from 'react';
import { ApplicationProvider } from '../../../contexts/application';
import { MemoryRouter } from 'react-router-dom';
import FoiActions from '../foi-actions';


const page = {
    params: {
        caseId: 'some_case_id',
        stageId: 'some_stage_id',
    }
};


describe('The FoiActions component', () => {
    it('should show the extend option when no extension applied', () => {
        const summary = {
            case: {
                deadline: '05/07/2021',
                deadLineExtensions: []
            }
        };

        const config = {
            page,
            summary
        };
        expect(
            render(<ApplicationProvider config={config}>
                <MemoryRouter>
                    <FoiActions />
                </MemoryRouter>
            </ApplicationProvider>)
        ).toMatchSnapshot();
    });

    it('should hide the extend option when FOI_PIT extension applied', () => {
        const summary = {
            case: {
                deadline: '05/07/2021',
                deadLineExtensions: { 'FOI_PIT':20 }
            }
        };

        const config = {
            page,
            summary
        };
        expect(
            render(<ApplicationProvider config={config}>
                <MemoryRouter>
                    <FoiActions />
                </MemoryRouter>
            </ApplicationProvider>)
        ).toMatchSnapshot();
    });
});
