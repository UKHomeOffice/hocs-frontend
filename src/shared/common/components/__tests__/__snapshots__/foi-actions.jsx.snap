// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The FoiActions component should hide the extend option when FOI_PIT extension applied 1`] = `
Array [
  <h2
    class="govuk-heading-m"
  >
    Case actions
  </h2>,
  <h3
    class="govuk-heading-s"
  >
    Extensions
  </h3>,
  <span
    class="govuk-body full-width"
  >
    Current due date: 05/07/2021 
    <span
      class="govuk-body"
    >
      (PIT Extension Applied)
    </span>
  </span>,
  <p
    class="govuk-body"
  >
    Apply an extension to this case.
  </p>,
  <hr
    class="govuk-section-break govuk-section-break--m govuk-section-break--visible"
  />,
  <h3
    class="govuk-heading-s"
  >
    Appeals
  </h3>,
  <p
    class="govuk-body"
  >
    Record an appeal against this request.
  </p>,
]
`;

exports[`The FoiActions component should show the extend option when no extension applied 1`] = `
Array [
  <h2
    class="govuk-heading-m"
  >
    Case actions
  </h2>,
  <h3
    class="govuk-heading-s"
  >
    Extensions
  </h3>,
  <span
    class="govuk-body full-width"
  >
    Current due date: 05/07/2021 
  </span>,
  <p
    class="govuk-body"
  >
    Apply an extension to this case.
  </p>,
  <a
    class="govuk-body govuk-link"
    href="/case/some_case_id/stage/some_stage_id/entity/actions_tab/extend_foi_deadline"
  >
    Extend this case
  </a>,
  <hr
    class="govuk-section-break govuk-section-break--m govuk-section-break--visible"
  />,
  <h3
    class="govuk-heading-s"
  >
    Appeals
  </h3>,
  <p
    class="govuk-body"
  >
    Record an appeal against this request.
  </p>,
]
`;
